### DISCUSSION TITLE
Troubles with the Maxwell solver ('yee')

### DISCUSSION THREAD

**Discussion by arka378** [epochpic/epoch#641] [General]
Hi,

I have been using EPOCH recently for my simulations but ran into some difficulty regarding the Maxwell solver.  

I am using the 'yee' solver. To test this solver, I created a setup using the latest EPOCH (EPOCH2D version 4.19.3   (commit v4.19.3-0-g724471f-clean)) for a EM wave, polarized along y and travelling along x. This simulation has no particles. The grid is 2D, with dimensions of 1 wavelength in x and 10 wavelengths in y. 

I evolved this setup for 1 waveperiod and measured the error in the electromagnetic fields, compared to the initial fields. I expected a secon-order behaviour from setting field_order=2, 4th order behaviour from setting field_order=4 and so on, by varying the resolution. But I get a strange behaviour, see figures attached.

For my simulation, I need the fields to be stable for about 2000 waveperiods. But with such behaviour with resolution, I cannot develop my simulation. How can I get better behaviour from the Maxwell solver, i.e better behaviour with resolution and higher accuracy over longer duration?

![fo2](https://github.com/Warwick-Plasma/epoch/assets/42378671/b304f2a3-8ce3-4cfa-a6f3-482c97385fcb)
![fo4](https://github.com/Warwick-Plasma/epoch/assets/42378671/283b86ad-7c61-4859-bd75-4081f6c635ed)
![fo6](https://github.com/Warwick-Plasma/epoch/assets/42378671/1ad347f0-7957-4f55-b150-883c1bb4e50a)

**Comment by Status-Mirror** [comment]
Hi @arka378,

I don't fully understand the plots you've attached. Could you provide an example input deck, and tell me output you're expecting?

Cheers,
Stuart

**Comment by arka378** [comment]
Hey @Status-Mirror!

I am expecting, for example, the `Yee` solver to be 2nd order when `field_order=2` in the `control` block. This means when I change the resolution 2x, errors should go down by 4x. From the first plot, you can see this is not the case (the log-log slope is only 1.37). I get worse behavior for higher orders.

The x-axis in the figures is inverse resolution. Moving to the left implies increasing resolution.
The y-axis in the figures is the relative error in the fields, after 1 wave period.

An example input deck is attached in a zip.

Best,
Arka

[mwe.deck.zip](https://github.com/Warwick-Plasma/epoch/files/14478343/mwe.deck.zip)

**Comment by Status-Mirror** [comment]
Hey Arka,

The EPOCH field-solver calculates Maxwell's equations, which involve spatial derivatives of field quantities. When you increase the field order, you increase the number of cells used to calculate these gradients - hopefully smoothing out any local fluctuations. No corrections are made to the laser or outflow boundaries as far as I can see.

With this in mind, I'm under the impression field-order is acting as a smoothing function. I'm not sure you should expect the 2x and 4x error reduction here. An increased field-order may help in some problems, but you may find it unhelpful in others.

I should also note that I wasn't personally involved in coding up the higher-order Yee field-solvers.

Cheers,
Stuart
